File: opcua-webservice-spring/examples/uawebservicetest/src/main/java/org/opcfoundation/uawebservicetest/testdigitaltwin/ElementListTestType.java
package org.opcfoundation.uawebservicetest.testdigitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaDataTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.element.ElementListType;
import org.opcfoundation.webserver.service.message.digitaltwin.*;
import org.opcfoundation.webserver.types.common.UaChildId;
import org.opcfoundation.webserver.types.digitaltwin.ObjectElementDescriptor;
import org.opcfoundation.webserver.types.digitaltwin.PropertyElementDescriptor;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public class ElementListTestType extends ElementListType {
    public ElementListTestType(DigitalTwinSpace space) {
        super("ElementListTestType", new LocalizedText("ElementListTestType"), space);
        setDescription(new LocalizedText("ElementListTestType"));
    }

    @Override
    public CompletableFuture<GetObjectElementListResponse> onGetObjectElementList(GetObjectElementListRequest request)
    {
        GetObjectElementListResponse response = new GetObjectElementListResponse();

        for (int i=request.getOffset();i<request.getOffset()+5; ++i)
        {
            response.add(new ObjectElementDescriptor(Integer.toString(i), new LocalizedText(Integer.toString(i)), DigitalTwinSpaceTest.elementCollectionTestAType));
        }

        response.setContainsMoreData(request.getOffset() == 0);

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public boolean supportObjectElementList() {
        return true;
    }

    @Override
    public CompletableFuture<GetPropertyElementListResponse> onGetPropertyElementList(GetPropertyElementListRequest request) {
        GetPropertyElementListResponse response = new GetPropertyElementListResponse();

        for (int i=request.getOffset();i<request.getOffset()+5; ++i)
        {
            response.add(new PropertyElementDescriptor(Integer.toString(i), new LocalizedText("Var-"+Integer.toString(i)), UaVariableTypes.BaseAnalogType));
        }

        response.setContainsMoreData(request.getOffset() == 0);

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public boolean supportPropertyElementList() {
        return true;
    }

    @Override
    public CompletableFuture<GetPropertyDescriptorResponse> onGetPropertyDescriptor(GetPropertyDescriptorRequest request) {
        GetPropertyDescriptorResponse response;

        if (null == request.getSubElementName())
        {
            response = new GetPropertyDescriptorResponse(
                new LocalizedText("Var-" + request.getPropertyId()),
                new LocalizedText("Variable number " + request.getPropertyId()),
                UaDataTypes.String,
                true);
        } else {
            response = new GetPropertyDescriptorResponse(
                    new LocalizedText("Element-" + request.getSubElementName()),
                    new LocalizedText("Sub-element " + request.getSubElementName()),
                    UaDataTypes.String,
                    false);
        }

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<ReadPropertyListValueResponse> onReadPropertyValues(ReadPropertyListValueRequest request) {
        ReadPropertyListValueResponse response = new ReadPropertyListValueResponse();

        for (String item: request.getPropertyIds())
        {
            response.setValue(item, Variant.ofString(item));
        }

        for (UaChildId item: request.getSubPropertyIds())
        {
            if (null == item.getSubElementName()) continue;
            response.setValue(item, Variant.ofString(item.getSubElementName()));
        }

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<WritePropertyListValuesResponse> onWritePropertyValues(WritePropertyListValuesRequest request) {
        System.out.println("Write object, " + request.getId());

        WritePropertyListValuesResponse response = new WritePropertyListValuesResponse();

        for (Map.Entry<String, Variant> item: request.getPropertyIdsAndValues().entrySet())
        {
            System.out.println(item);
            response.setWriteValueResult(item.getKey(), StatusCode.GOOD);
        }

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<GetPropertySubElementsResponse> onGetPropertySubElements(GetPropertySubElementsRequest request) {
        GetPropertySubElementsResponse response = new GetPropertySubElementsResponse();
        response.add("Sub-" + request.getPropertyId());
        return CompletableFuture.completedFuture(response);
    }
}

File: opcua-webservice-node/examples/testserver/testdigitaltwin/ElementListTestType.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaLocalizedText, UaVariant } from "opcua-webapi-ts";
import { UaDataTypes, UaVariableTypes } from "../../../src/server/addressspace/nodes";
import { DigitalTwinSpace } from "../../../src/server/digitaltwin";
import { ElementListType } from "../../../src/server/digitaltwin/element";
import {
    GetObjectElementListRequest,
    GetObjectElementListResponse,
    GetPropertyDescriptorRequest,
    GetPropertyDescriptorResponse,
    GetPropertyElementListRequest,
    GetPropertyElementListResponse,
    GetPropertySubElementsRequest,
    GetPropertySubElementsResponse,
    ReadPropertyListValueRequest,
    ReadPropertyListValueResponse,
    WritePropertyListValuesRequest,
    WritePropertyListValuesResponse,
} from "../../../src/server/service/message";
import { ObjectElementDescriptor } from "../../../src/server/types/digitaltwin/ObjectElementDescriptor";
import { PropertyElementDescriptor } from "../../../src/server/types/digitaltwin/PropertyElementDescriptor";
import { DigitalTwinSpaceTest } from "./DigitalTwinSpaceTest";

export class ElementListTestType extends ElementListType {
    constructor(space: DigitalTwinSpace) {
        super("ElementListTestType", new UaLocalizedText("ElementListTestType"), space);
        this.description = new UaLocalizedText("ElementListTestType");
    }

    override async onGetObjectElementList(request: GetObjectElementListRequest): Promise<GetObjectElementListResponse> {
        const response = new GetObjectElementListResponse();

        for (let index = request.offset; index < request.offset + 5; ++index) {
            response.add(
                new ObjectElementDescriptor(
                    `${index}`,
                    new UaLocalizedText(`${index}`),
                    DigitalTwinSpaceTest.elementCollectionTestAType,
                ),
            );
        }

        response.containsMoreData = request.offset === 0;
        return response;
    }

    override supportObjectElementList(): boolean {
        return true;
    }

    override async onGetPropertyElementList(request: GetPropertyElementListRequest): Promise<GetPropertyElementListResponse> {
        const response = new GetPropertyElementListResponse();

        for (let index = request.offset; index < request.offset + 5; ++index) {
            response.add(
                new PropertyElementDescriptor(
                    `${index}`,
                    new UaLocalizedText(`Var-${index}`),
                    UaVariableTypes.DataItemType,
                ),
            );
        }

        response.containsMoreData = request.offset === 0;
        return response;
    }

    override supportPropertyElementList(): boolean {
        return true;
    }

    override async onGetPropertyDescriptor(request: GetPropertyDescriptorRequest): Promise<GetPropertyDescriptorResponse> {
        if (request.subElementName === null) {
            return new GetPropertyDescriptorResponse(
                new UaLocalizedText(`Var-${request.propertyId}`),
                new UaLocalizedText(`Variable number ${request.propertyId}`),
                UaDataTypes.String,
                true,
            );
        }

        return new GetPropertyDescriptorResponse(
            new UaLocalizedText(`Element-${request.subElementName}`),
            new UaLocalizedText(`Sub-element ${request.subElementName}`),
            UaDataTypes.String,
            false,
        );
    }

    override async onReadPropertyValues(request: ReadPropertyListValueRequest): Promise<ReadPropertyListValueResponse> {
        const response = new ReadPropertyListValueResponse();

        for (const propertyId of request.propertyIds) {
            response.setValue(propertyId, UaVariant.string(propertyId));
        }

        for (const subPropertyId of request.subPropertyIds) {
            if (subPropertyId.subElementName === null) {
                continue;
            }

            response.setValue(subPropertyId, UaVariant.string(subPropertyId.subElementName));
        }

        return response;
    }

    override async onWritePropertyValues(request: WritePropertyListValuesRequest): Promise<WritePropertyListValuesResponse> {
        console.log(`Write object, ${request.id}`);

        const response = new WritePropertyListValuesResponse();

        for (const [key, value] of request.propertyIdsAndValues) {
            console.log(key, value.value);
            response.setWriteValueResult(key, makeUaStatusCode(StatusCodes.Good));
        }

        for (const [key, value] of request.subPropertyIdsAndValues) {
            console.log(key.toString(), value.value);
            response.setWriteValueResult(key, makeUaStatusCode(StatusCodes.Good));
        }

        return response;
    }

    override async onGetPropertySubElements(request: GetPropertySubElementsRequest): Promise<GetPropertySubElementsResponse> {
        const response = new GetPropertySubElementsResponse();
        response.add(`Sub-${request.propertyId}`);
        return response;
    }
}